Clone Snapshot API

original-brownbear opened the pull request:
Snapshot clone API. Complete except for some TODOs around documentation (and adding HLRC support).

original-brownbear committed:
bck

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
Clone snapshot API start

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
works

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
r/w

original-brownbear committed:
bck

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
fix

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
progress

original-brownbear committed:
bck

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
fix

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
bck

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
bck

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
bck

original-brownbear committed:
bck

original-brownbear committed:
bck

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
bck

original-brownbear committed:
bck

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
bck

original-brownbear committed:
worksish

original-brownbear committed:
bck

original-brownbear committed:
bck

original-brownbear committed:
bck

original-brownbear committed:
bck

original-brownbear committed:
better

original-brownbear committed:
bck

original-brownbear committed:
green

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
bck

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
bck

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
clone prevents delete

original-brownbear committed:
delete prevents clone

original-brownbear committed:
bck

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
bck

original-brownbear committed:
cleanup rest action

original-brownbear committed:
shorter diff

original-brownbear committed:
bck

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
docs test

original-brownbear committed:
one more case handled

original-brownbear committed:
error handling

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
more fixes

original-brownbear committed:
last one?

original-brownbear committed:
bck

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
fix state machine

original-brownbear committed:
optimize

original-brownbear committed:
better not great

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
mroe efficient

original-brownbear committed:
fix

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
fix

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
Merge branch 'master' of github.com:elastic/elasticsearch into clone-snapshot-on-repo

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
reproducer

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
fix master failover part 1

original-brownbear committed:
fix master failover part 2

original-brownbear committed:
nicer

original-brownbear committed:
nicer

original-brownbear committed:
cleaner

original-brownbear committed:
fix more edge cases

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
shorter

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
fix docs?

original-brownbear committed:
remove settings for now

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
less noise

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
remove noisy changes

original-brownbear committed:
reduce noise

original-brownbear committed:
align style

original-brownbear committed:
align style

original-brownbear committed:
making things look nicer

original-brownbear committed:
drier

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
drier

original-brownbear committed:
nicer

original-brownbear committed:
make some things nicer

original-brownbear committed:
add validation

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
less noisy

original-brownbear committed:
less noise

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
cleaner

original-brownbear committed:
safer

original-brownbear committed:
better docs

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
better

original-brownbear committed:
better

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
nicer

original-brownbear committed:
simpler diff

original-brownbear committed:
fix

original-brownbear committed:
TODO

original-brownbear committed:
tests are hard

elasticmachine commented:
Pinging @elastic/es-distributed (:Distributed/Snapshot/Restore)

tlrx commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -313,6 +308,275 @@ public TimeValue timeout() {
+        final String snapshotName = indexNameExpressionResolver.resolveDateMathExpression(request.target());
+        validate(repositoryName, snapshotName);
+        Repository repository = repositoriesService.repository(repositoryName);
+        if (repository.isReadOnly()) {
```
Maybe move this few steps up, after `final String repositoryName = request.repository();`

original-brownbear replied:
done :)

tlrx commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -313,6 +308,275 @@ public TimeValue timeout() {
+        final SnapshotId targetSnapshot = target.getSnapshotId();
+        final String localNodeId = clusterService.localNode().getId();
+        if (currentlyCloning.add(repoShardId)) {
+            repository.cloneShardSnapshot(sourceSnapshot, targetSnapshot, repoShardId,
```
nit: indentation messed up here, making the code difficult to read

tlrx commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1159,6 +1465,17 @@ public ClusterState execute(ClusterState currentState) {
+                        .stream()
+                        .filter(entry -> entry.source() != null)
+                        .map(SnapshotsInProgress.Entry::source
+                        ).collect(Collectors.toSet());
```
nit: ident

tlrx commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1809,16 +2126,32 @@ public static ClusterState updateWithSnapshots(ClusterState state,
-            for (ObjectObjectCursor<ShardId, ShardSnapshotStatus> shard : runningSnapshot.shards()) {
-                if (shard.value.isActive()) {
-                    inProgressShards.add(shard.key);
+            if (runningSnapshot.source() == null) {
```
It's use everywhere, maybe adding a method `isClone()` returning `source != null` would be more explicit?

original-brownbear replied:
++  added that, much nicer now :)

tlrx commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1809,16 +2126,32 @@ public static ClusterState updateWithSnapshots(ClusterState state,
      * @return shard ids that currently have an actively executing shard snapshot on a data node
      */
-    private static Set<ShardId> busyShardsForRepo(String repoName, @Nullable SnapshotsInProgress snapshots) {
+    private static Set<ShardId> busyShardsForRepo(String repoName, @Nullable SnapshotsInProgress snapshots, Metadata metadata) {
```
Maybe `activeShardsForRepo` would be more appropriate?

original-brownbear replied:
Maybe I'm indifferent to the naming here, I'd say let's not rename things we don't have to rename in this one to keep the noise down maybe? There's a TODO to refactor the use of this method anyway and we can do better with the naming when we get to that?

tlrx reviewed (commented):
I made a first pass, mostly stupid cosmetic suggestions. I'm still digesting the changes but will need more time.

original-brownbear commented on server/src/internalClusterTest/java/org/elasticsearch/snapshots/CloneSnapshotIT.java:
```diff
@@ -107,6 +122,365 @@ public void testShardClone() throws Exception {
+        assertAllSnapshotsSuccessful(getRepositoryData(repoName), cloneSucceeded ? 2 : 1);
+    }
+
+    public void testFailsOnCloneMissingIndices() {
```
Yea we definitely should have one (could've sworn I added one already but apparently not ...), on it.

original-brownbear replied:
Added in 6e5b74bfcd4  :)

original-brownbear commented on server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java:
```diff
@@ -96,18 +99,52 @@ public static Entry startedEntry(Snapshot snapshot, boolean includeGlobalState,
+     */
+    public static Entry startClone(Snapshot snapshot, SnapshotId source, List<IndexId> indices, long startTime,
+                                   long repositoryStateId, Version version) {
+        return new SnapshotsInProgress.Entry(snapshot, false, false, State.STARTED, indices, Collections.emptyList(),
```
My understanding would be yes since the motivation here was cloning indices. That said, I could just add the ability to clone global meta as well in a follow up? It's not hard to add but adds yet another step of complexity to snapshot finalization so that going into a separate step might be a good idea if we want it?

tlrx replied:
Again I'm afraid of silent loses. If a snapshot contains the global metadata I'd expect its clone to also contains it (or vice versa). I'm ok for a follow up as long as it lands in the feature before it's backported.

original-brownbear replied:
Actually ... turns out bdd6943f913 is all that's needed to get the global metadata into the clone. The more interesting question is data streams. I'm not sure we would just want to clone those as well every time because of how they interact with indices. That might technically need smarter logic there, I'll look into those in a follow-up.

original-brownbear committed:
fix doc

original-brownbear committed:
fix doc

original-brownbear commented on docs/reference/snapshot-restore/take-snapshot.asciidoc:
```diff
@@ -124,3 +124,5 @@ PUT /_snapshot/my_backup/<snapshot-{now/d}>
 -----------------------------------
 // TEST[continued]
+
+NOTE: You can also create snapshots that are copies of part of an existing snapshot using the <<clone-snapshot-api,clone snapshot API>>.
```
This was actually intentional. Cloning is there to copy part of a snapshot, cloning a complete snapshot doesn't really make any sense.

original-brownbear committed:
include global metadata

original-brownbear committed:
fix indent

original-brownbear committed:
fix indent

tlrx commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1911,82 +2244,246 @@ public boolean assertAllListenersResolved() {
+                        // start a shard snapshot or clone operation on the current entry
+                        if (entry.source() == null) {
+                            // current entry is a snapshot operation so we must translate the repository shard id to a routing shard id
+                            final IndexMetadata indexMeta = currentState.metadata().index(finishedShardId.indexName());
```
I've been a bit worried about re-resolving indices here as the cloned shard could reference an index from the snapshot that do not exist anymore, or could have been recreated with a different sharding, but after writing some tests I think it's OK as `existingStatus == null` later will prevent any difference in the number of shards.

original-brownbear committed:
add assertion

original-brownbear commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1911,82 +2244,246 @@ public boolean assertAllListenersResolved() {
+                    // The update applied to a shard clone operation
+                    final RepositoryShardId finishedShardId = updateSnapshotState.repoShardId;
+                    if (entry.snapshot().getSnapshotId().equals(updatedSnapshot.getSnapshotId())) {
+                        final ShardSnapshotStatus existing = entry.clones().get(finishedShardId);
```
certainly can, added it :)

original-brownbear committed:
isCLone

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
some UTs

original-brownbear committed:
moar UTs

original-brownbear committed:
even moar tests

original-brownbear committed:
even moar tests (80% cov)

original-brownbear committed:
full coverage

original-brownbear committed:
shorter

original-brownbear commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1911,82 +2244,246 @@ public boolean assertAllListenersResolved() {
+     * will be started for that snapshot entry and the task removed from the collection of tasks that need to be applied to snapshot
+     * entries since it can not have any further effects.
+     */
     private static final ClusterStateTaskExecutor<ShardSnapshotUpdate> SHARD_STATE_EXECUTOR = (currentState, tasks) -> {
```
Added a few of those now with effectively 100% coverage :)

original-brownbear committed:
IT

original-brownbear commented:
Alrighty @tlrx thanks so much for working through this one! I addressed the last round of comments now and added UTs and and IT for a broken source snapshot as request. Should be good for another look :)

tlrx commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -313,6 +308,276 @@ public TimeValue timeout() {
+            for (IndexId indexId : indices) {
+                if (RestoreService.failed(snapshotInfo, indexId.getName())) {
+                    throw new SnapshotException(targetSnapshot, "Can't clone index [" + indexId +
+                            "] because its snapshot is was not successful.");
```
```suggestion
                            "] because its snapshot was not successful.");
```

tlrx reviewed (approved):
Thanks for the additional tests, they worth it IMO. Not funny to write but I think it helps to reason about the shard snapshot updating method.

LGTM

original-brownbear committed:
Update server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

Co-authored-by: Tanguy Leroux <tlrx.dev@gmail.com>

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into clone-snapshot-on-repo

original-brownbear committed:
gotta fix message in tests as well

original-brownbear commented:
Thanks again Tanguy!
Ran this for about 5h at 8x parallelism through all snapshot related test on my local workstation without failure => merging and on it for the last few TODOs :)
